### PR TITLE
Road to configV6

### PR DESCRIPTION
**Pull request by z4kn4fein**
Implements config v6 support.

Features/improvements:
* [x] Fetch new config json version (`config_v6.json`) and update config model
* [x] SDK key format validation (when client is not set up to use local-only flag overrides)
* [x] AND conditions
* [x] New condition types
   * Prerequisite flag condition
   * Segment condition
* [x] Changed comparison operators
   * (NOT) CONTAINS ANY OF
* [x] New comparison operators
   * BEFORE/AFTER
   * TEXT (NOT) EQUALS (both hashed and cleartext)
   * TEXT (NOT) STARTS/ENDS WITH ANY OF (both hashed and cleartext)
   * ARRAY CONTAINS ANY OF (both hashed and cleartext)
* [x] Percentage options within targeting rules
* [x] Percentage options based on custom attributes
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] User Object attribute/comparison value trimming
* [x] Comparison value hashing changes (SHA256, salting)
* [ ] ~Flag overrides vs. config json salt and segments~
* [x] `EvaluationDetails.getMatchedTargetingRule`/`getMatchedPercentageOption` properties (rename + set combinations correctly)
* [x] Update `ConfigCatClient.getKeyAndValue`

Tests: 
* [x] Update cache key/payload tests
* [x] Evaluation matrix tests
  * Config v1 features (both V1 & V2)
  * Config v2 features
* [x] Evaluation log tests
* [x] SDK key format validation
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] Flag overrides vs. prerequisite flags (circular dependency, comparison value type mismatch, dependent/prereq flag overriding)
* [ ] ~Flag overrides vs. config json salt and segments~
* [x] `EvaluationDetails.getMatchedTargetingRule`/`getMatchedPercentageOption` properties (rename + set combinations correctly)
* [x] Special character tests (https://trello.com/c/qh4jHrvm)
* [x] User attribute / comparison value trim tests (https://trello.com/c/vt8FFrSZ)
* [x] Comparison attribute and comparison value trimming tests (https://trello.com/c/vt8FFrSZ)
* [x] Comparison attribute conversion to canonical string representation tests (https://trello.com/c/CZgGrrcb)

Related improvements/fixes:
* [ ] ~Dynamically typed languages should log a warning when default value doesn't match the setting type (see https://trello.com/c/Le6vimGu)~
* [x] We should go to the cache in all polling modes instead of using the in memory variable (see https://trello.com/c/rreKm64A)
* [x] Improve message of log event 1103 (https://trello.com/c/Mb7rvm64)

Additional improvements:
- Ability to pass a custom logger (by conforming `protocol ConfigCatLogger`) implementation with the `logger` configuration option. The default logger implementation still using `os_log`.
- [Privacy manifest](https://developer.apple.com/documentation/bundleresources/privacy_manifest_files#4284009) file for the project.
- `visionOS` support.

Breaking changes:
- The `onConfigChanged` hook now gets a `Config` object that contains all the newly downloaded settings & segments.
- The `custom` init parameter of `ConfigCatUser` now accepts `Any` typed values not just strings. This allows the usage of various types in the user object's custom attributes.
- Renamed `matchedEvaluationRule` / `matchedEvaluationPercentageRule` properties of `EvaluationDetails` to `matchedTargetingRule` / `matchedPercentageOption`
- Renamed `LogLevel` to `ConfigCatLogLevel`.

### Requirement checklist (only if applicable)

- [x] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

### CODE REVIEW

**Review by adams85 — CHANGES_REQUESTED** [review]
Besides the comments,
* Although the checkbox for [this test](https://github.com/configcat/.net-sdk/blob/4692622a4cb54fce272b3689f6d13dee3be8e96b/src/ConfigCat.Client.Tests/BasicConfigCatClientIntegrationTests.cs#L464-L477) is checked, I couldn't find it.
* The error message of `FlagEvaluator.validateFlagType` is not consistent with other SDKs. See https://github.com/configcat/.net-sdk/blob/4692622a4cb54fce272b3689f6d13dee3be8e96b/src/ConfigCatClient/Extensions/TypeExtensions.cs#L11

File: Sources/ConfigCat/Config.swift [hunk 1]
```diff
@@ -77,226 +278,726 @@ class ConfigEntry: Equatable {
...
+    static let percentageOptionsKey = "p"
+
+    /// The value associated with the targeting rule or nil if the targeting rule has percentage options THEN part.
+    @objc public let servedValue: ServedValue?
+
+    /// The list of conditions that are combined with the AND logical operator.
+    @objc public let conditions: [Condition]
+
+    /// The list of percentage options associated with the targeting rule or nil if the targeting rule has a served value THEN part.
+    @objc public let percentageOptions: [PercentageOption]
```
Comment (adams85) [inline Sources/ConfigCat/Config.swift:511]:
```suggestion
    /// The list of percentage options associated with the targeting rule or empty if the targeting rule has a served value THEN part.
    @objc public let percentageOptions: [PercentageOption]
```

File: Sources/ConfigCat/Config.swift [hunk 2]
```diff
@@ -77,226 +278,726 @@ class ConfigEntry: Equatable {
...
+    @objc public let stringValue: String?
+    
+    /// Value in numeric format that the User Object attribute is compared to.
+    public let doubleValue: Double?
+    
+    /// Value in numeric Objective-C format that the User Object attribute is compared to.
+    @objc public let doubleValueObjC: NSNumber?
+    
+    /// Value in text array format that the User Object attribute is compared to.
+    @objc public let stringArrayValue: [String]?
```
Comment (adams85) [inline Sources/ConfigCat/Config.swift:616]:
```suggestion
    /// The value that the User Object attribute is compared to, when the comparator works with a single text comparison value.
    @objc public let stringValue: String?
    
    /// The value that the User Object attribute is compared to, when the comparator works with a numeric comparison value.
    public let doubleValue: Double?
    
    /// The value in Objective-C format that the User Object attribute is compared to, when the comparator works with a numeric comparison value.
    @objc public let doubleValueObjC: NSNumber?
    
    /// The value that the User Object attribute is compared to, when the comparator works with an array of text comparison values.
    @objc public let stringArrayValue: [String]?
```

File: Sources/ConfigCat/Config.swift [hunk 3]
```diff
@@ -77,226 +278,726 @@ class ConfigEntry: Equatable {
...
+
+public final class ServedValue: NSObject, JsonSerializable {
+    static let valueKey = "v"
+    static let idKey = "i"
+
+    /// The value associated with the targeting rule. It's empty (its `isEmpty` property is `true`) when the targeting rule has percentage options THEN part.
+    @objc public let value: SettingValue
+    
+    /// Variation ID.
+    @objc public let variationId: String?
```
Comment (adams85) [inline Sources/ConfigCat/Config.swift:941]:
```suggestion
    /// The value associated with the targeting rule.
    @objc public let value: SettingValue
    
    /// Variation ID (for analytical purposes).
    @objc public let variationId: String?
```

File: Sources/ConfigCat/Config.swift [hunk 4]
```diff
@@ -77,226 +278,726 @@ class ConfigEntry: Equatable {
...
 
-    /// Value served when the rule is selected during evaluation.
-    @objc public let value: Any
+    /// The served value of the percentage option.
+    @objc public let servedValue: SettingValue
 
     /// The rule's percentage value.
     @objc public let percentage: Int
 
     /// The rule's variation ID (for analytical purposes).
```
Comment (adams85) [inline Sources/ConfigCat/Config.swift:973]:
```suggestion
    /// A number between 0 and 100 that represents a randomly allocated fraction of the users.
    @objc public let percentage: Int

    /// Variation ID (for analytical purposes).
```

File: Sources/ConfigCat/Config.swift [hunk 5]
```diff
@@ -1,5 +1,206 @@
...
+            return "Int"
+        case .double:
+            return "Double"
+        }
+    }
+}
+
+@objc public enum UserComparator: Int {
+    case unknown = -1
+    /// Checks whether the comparison attribute is equal to any of the comparison values.
```
Comment (adams85) [inline Sources/ConfigCat/Config.swift:63]: ❤️

File: Sources/ConfigCat/ConfigCatUser.swift [hunk 6]
```diff
@@ -1,60 +1,98 @@
...
+     - Parameter country: Optional, country of the user.
+     - Parameter custom: Optional, custom attributes of the user for advanced targeting rule definitions (e.g. role, subscription type, etc.)
+     - Returns: A new `ConfigCatUser`.
+     
+     All comparators support `String` values as User Object attribute (in some cases they need to be provided in a specific format though, see below),
+     but some of them also support other types of values. It depends on the comparator how the values will be handled. The following rules apply:
+     
+     **Text-based comparators** (EQUALS, IS ONE OF, etc.)
+     * accept `String` values,
+     * all other values are automatically converted to `string` (a warning will be logged but evaluation will continue as normal).
```
Comment (adams85) [inline Sources/ConfigCat/ConfigCatUser.swift:26]:
```suggestion
     * all other values are automatically converted to `String` (a warning will be logged but evaluation will continue as normal).
```

File: Sources/ConfigCat/ConfigCatUser.swift [hunk 7]
```diff
@@ -1,60 +1,98 @@
...
+     * all other values are considered invalid (a warning will be logged and the currently evaluated targeting rule will be skipped).
      
-     - Parameter identifier: the SDK Key for to communicate with the ConfigCat services.
-     - Parameter email: optional, sets the email of the user.
-     - Parameter country: optional, sets the country of the user.
-     - Parameter custom: optional, sets the custom attributes of a user.
-     - Returns: A new `User`.
+     **Number-based comparators** (=, &lt;, &gt;=, etc.)
+     * accept `Int`, `UInt`, `Double`, or `Float` values,
+     * accept `String` values containing a properly formatted, valid `Int`, `UInt`, `Double`, or `Float` value,
```
Comment (adams85) [inline Sources/ConfigCat/ConfigCatUser.swift:34]:
```suggestion
     * accept `String` values containing a properly formatted, valid `Double` value,
```

File: Sources/ConfigCat/ConfigCatUser.swift [hunk 8]
```diff
@@ -1,60 +1,98 @@
...
+     * all other values are considered invalid (a warning will be logged and the currently evaluated targeting rule will be skipped).
+     
+     **Date time-based comparators** (BEFORE / AFTER)
+     * accept `Date` values, which are automatically converted to a second-based Unix timestamp,
+     * accept `Int`, `UInt`, `Double`, or `Float` values representing a second-based Unix timestamp,
+     * accept `String` values containing a properly formatted, valid `Int`, `UInt`, `Double`, or `Float` value,
```
Comment (adams85) [inline Sources/ConfigCat/ConfigCatUser.swift:40]:
```suggestion
     * accept `String` values containing a properly formatted, valid `Double` value,
```

**Review by adams85 — CHANGES_REQUESTED** [review]
Another (hopefully last) batch of minor change requests.

Besides these, please
* [ ] Add the warning about allowed generic arguments ("The generic parameter `Value` represents...") to `ConfigCatClient.getValue` and `getValueDetails` as well.

File: Sources/ConfigCat/RolloutEvaluator.swift [hunk 9]
```diff
@@ -6,323 +6,809 @@ import os.log
...
+            guard let compVal = cond.stringValue else {
+                return .compValueInvalid(nil)
+            }
+            let (userVal, converted) = asString(value: userAnyVal)
+            if converted {
+                logConverted(cond: cond, key: key, attrValue: userVal)
+            }
+            guard let salt = salt else {
+                return .fatal(RolloutEvaluator.saltMissingMessage)
+            }
```
[the remark this answers is not in the document]
Comment (z4kn4fein) [reply Sources/ConfigCat/RolloutEvaluator.swift]: Done

File: Sources/ConfigCat/RolloutEvaluator.swift [hunk 10]
```diff
@@ -6,323 +6,809 @@ import os.log
...
+        let needsTrue = comp.isStartsWith ? comp.isSensitive ? comp == .startsWithAnyOfHashed : comp == .startsWithAnyOf :
+        comp.isSensitive ? comp == .endsWithAnyOfHashed : comp == .endsWithAnyOf
+        let userValData = Data(userValue.utf8)
+        for value in comparisonValue {
+            if comp.isSensitive {
+                let parts = value.components(separatedBy: "_")
+                if parts.count < 2 || parts[1].isEmpty {
+                    return .fatal("Comparison value is missing or invalid")
+                }
+                guard let length = Int(parts[0].trimmingCharacters(in: .whitespaces)), length > 0 else {
```
[the remark this answers is not in the document]
Comment (z4kn4fein) [reply Sources/ConfigCat/RolloutEvaluator.swift]: Done

File: Sources/ConfigCat/EvaluationLogger.swift [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        } else {
+            self.append(value: " ")
+        }
+        self.append(value: "THEN")
+        
+        if let val = targetingRule.servedValue?.value.anyValue {
+            self.append(value: " '\(val)'")
+        } else if !targetingRule.percentageOptions.isEmpty {
+            self.append(value: " % options")
+        }
```
[the remark this answers is not in the document]
Comment (z4kn4fein) [reply Sources/ConfigCat/EvaluationLogger.swift]: Done

File: Sources/ConfigCat/Config.swift [hunk 12]
```diff
@@ -77,226 +278,726 @@ class ConfigEntry: Equatable {
...
-    static let entriesKey = "f"
-
-    let preferences: Preferences?
-    let entries: [String: Setting]
+    static let settingsKey = "f"
+    static let segmentsKey = "s"
+
+    /// The dictionary of settings.
+    @objc public let settings: [String: Setting]
+    /// The dictionary of settings.
```
[the remark this answers is not in the document]
Comment (z4kn4fein) [reply Sources/ConfigCat/Config.swift]: Done

File: Sources/ConfigCat/Config.swift [hunk 13]
```diff
@@ -77,226 +278,726 @@ class ConfigEntry: Equatable {
...
-    static let rolloutRulesKey = "r"
+    static let percentageAttributeKey = "a"
+    static let settingTypeKey = "t"
+    static let percentageOptionsKey = "p"
+    static let targetingRulesKey = "r"
     static let variationIdKey = "i"
 
-    /// Value of the feature flag / setting.
-    @objc public let value: Any
+    /// The setting's default value used when no targeting rules are matching during an evaluation process.
```
[the remark this answers is not in the document]
Comment (z4kn4fein) [reply Sources/ConfigCat/Config.swift]: Done

File: Sources/ConfigCat/Config.swift [hunk 14]
```diff
@@ -77,226 +278,726 @@ class ConfigEntry: Equatable {
...
+    @objc public let targetingRules: [TargetingRule]
 
     /// Variation ID (for analytical purposes).
-    @objc public let variationId: String
+    @objc public let variationId: String?
+    
+    /// The User Object attribute which serves as the basis of percentage options evaluation.
+    @objc public let percentageAttribute: String
+    
+    /// The setting's type. It can be Bool, String, Int, Float.
```
[the remark this answers is not in the document]
Comment (z4kn4fein) [reply Sources/ConfigCat/Config.swift]: Done

File: Sources/ConfigCat/Config.swift (hunk 1 above)
Comment (z4kn4fein) [reply Sources/ConfigCat/Config.swift]: Done

File: Sources/ConfigCat/Config.swift (hunk 2 above)
Comment (z4kn4fein) [reply Sources/ConfigCat/Config.swift]: Done

File: Sources/ConfigCat/Config.swift (hunk 4 above)
Comment (z4kn4fein) [reply Sources/ConfigCat/Config.swift]: Done

File: Sources/ConfigCat/Config.swift (hunk 5 above)
Comment (z4kn4fein) [reply Sources/ConfigCat/Config.swift:63]: Don.... oops, almost!

File: Sources/ConfigCat/ConfigCatUser.swift (hunk 6 above)
Comment (z4kn4fein) [reply Sources/ConfigCat/ConfigCatUser.swift]: Done

File: Sources/ConfigCat/ConfigCatUser.swift (hunk 7 above)
Comment (z4kn4fein) [reply Sources/ConfigCat/ConfigCatUser.swift]: Done

File: Sources/ConfigCat/ConfigCatUser.swift (hunk 8 above)
Comment (z4kn4fein) [reply Sources/ConfigCat/ConfigCatUser.swift]: Done

**Comment by z4kn4fein** [comment]
> Another (hopefully last) batch of minor change requests.
> 
> Besides these, please
> 
> * [ ]  Add the warning about allowed generic arguments ("The generic parameter `Value` represents...") to `ConfigCatClient.getValue` and `getValueDetails` as well.

It's there on the client's functions: https://github.com/configcat/swift-sdk/blob/c748d1bab8fc64f6b11cba9f131e5fbd789ed3b9/Sources/ConfigCat/ConfigCatClient.swift#L165-L197
Did you mean somewhere else?

**Comment by adams85** [comment]
> Did you mean somewhere else?

Yep, I meant `ConfigCatClientSnapshot.getValue`, just couldn't type it out... :)

**Review by adams85 — APPROVED** [review]

**Pull request merged**